docs/pages/router/advanced/authentication-rewrites.mdx

```diff
@@ -0,0 +1,406 @@
+    initialValue
+  ) as UseStateHook<T>;
+}
+
+export async function setStorageItemAsync(key: string, value: string | null) {
+  if (Platform.OS === 'web') {
+    try {
+      if (value === null) {
+        localStorage.removeItem(key);
+      } else {
+        localStorage.setItem(key, value);
+      }
+    } catch (e) {
+      console.error('Local storage is unavailable:', e);
+    }
+  } else {
+    if (value == null) {
+      await SecureStore.deleteItemAsync(key);
+    } else {
+      await SecureStore.setItemAsync(key, value);
+    }
+  }
+}
+
+export function useStorageState(key: string): UseStateHook<string> {
+  // Public
+  const [state, setState] = useAsyncState<string>();
+
+  // Get
+  useEffect(() => {
+    if (Platform.OS === 'web') {
+      try {
+        if (typeof localStorage !== 'undefined') {
+          setState(localStorage.getItem(key));
+        }
+      } catch (e) {
+        console.error('Local storage is unavailable:', e);
+      }
+    } else {
+      SecureStore.getItemAsync(key).then(value => {
+        setState(value);
+      });
+    }
+  }, [key]);
+
+  // Set
+  const setValue = useCallback(
+    (value: string | null) => {
+      setState(value);
+      setStorageItemAsync(key, value);
+    },
+    [key]
+  );
+
+  return [state, setValue];
+}
+```
+
+</Collapsible>
+
+</Step>
+
+<Step label="2">
+
+Use the `SessionProvider` in the root layout to provide the authentication context to the entire app. It's imperative that the `<Slot />` is mounted before any navigation events are triggered. Otherwise, a runtime error will be thrown.
+
+```tsx app/_layout.tsx
+import { Slot } from 'expo-router';
+import { SessionProvider } from '../ctx';
+
+export default function Root() {
+  // Set up the auth context and render our layout inside of it.
+  return (
+    <SessionProvider>
+      <Slot />
+    </SessionProvider>
+  );
+}
+```
+
+</Step>
+
+<Step label="3">
+
+Create a nested [layout route](/router/basics/layout/) that checks whether users are authenticated before rendering the child route components. This layout route redirects users to the sign-in screen if they are not authenticated.
+
+```tsx app/(app)/_layout.tsx|collapseHeight=400
+import { Text } from 'react-native';
+import { Redirect, Stack } from 'expo-router';
+
+import { useSession } from '../../ctx';
+
+export default function AppLayout() {
+  const { session, isLoading } = useSession();
+
+  // You can keep the splash screen open, or render a loading screen like we do here.
+  if (isLoading) {
+    return <Text>Loading...</Text>;
+  }
+
+  // Only require authentication within the (app) group's layout as users
+  // need to be able to access the (auth) group and sign in again.
+  if (!session) {
+    // On web, static rendering will stop here as the user is not authenticated
+    // in the headless Node process that the pages are rendered in.
+    return <Redirect href="/sign-in" />;
```

Comment:
We should move away from returning `<Redirect />` from layout routes since it's `null` and not a layout.